ccmbioinfo/lfs
# CARDS-1423: Certain survey responses should trigger an immediate email to the on-call nurse

**Please note that PR uses `CARDS-1424` as a base and will be in _draft_ mode until `CARDS-1424` is merged into `dev`, at which point this PR's base will be switched to `dev`**.

This PR implements CARDS-1423 by providing an OSGi component factory for instantiating JCR listeners which will send emails to a clinic's emergency contact if a worrisome answer is given in a response to a questionnaire.

To test:

- Follow the instructions at https://github.com/data-team-uhn/cards/pull/822#issue-1063852994 to configure a local debugging mail server for CARDS to use.
- Start CARDS with the command `CARDS_HOST_AND_PORT='datapro.uhn.ca' CLINIC_SLING_PATH='/Proms.html/Cardio' PATIENT_NOTIFICATION_FROM_ADDRESS='datapro@uhn.ca' PATIENT_NOTIFICATION_FROM_NAME='UHN DATAPRO' SLING_COMMONS_CRYPTO_PASSWORD=password ./start_cards.sh --project cards4proms --dev -f mvn:io.uhndata.cards/cards-email-notifications/0.9-SNAPSHOT/slingosgifeature -V emailnotifications.smtps.port=8465 -V emailnotifications.smtps.host=localhost`.
- Login to CARDS as `admin`:`admin`.
- In a new tab, visit `http://localhost:8080/system/console/configMgr`, click the **`+`** button on _`Email Configured Feature`_. Use the default settings and add an optional _Alert Description_ (eg. _"This patient is at risk of harming themselves"_). Click _Save_. You may now close this browser tab.
- Go to _Administration_ ... _Questionnaires_ ... open _Visit information_ for editing.
- Open a new browser tab to `http://localhost:8080/bin/browser.html` and change the `displayMode` property on the `/Questionnaires/Visit information/surveys` node from `hidden` to `input`. Also, add the `emergencyContact` property with a value of `emergency@test.com` to the `/Proms/Cardio` JCR node. You may now close the `/bin/browser.html` tab.
- Create a new _Patient information_ Form and, with it, a new _Patient_ subject (`P1`). Complete the form as desired.
- Create a new _Visit information_ Form and, with it, a new _Visit_ subject (`V1`). Select `P1` as the _parent Patient_ for this `V1` subject.
- Set the value of the _Surveys_ field in this question to _Cardio_.
- Save the Form.
- Create a new _PHQ-9_ Form for `Patient P1 / Visit V1`
- Select _Nearly every day_ for the question _Feeling down, depressed, or hopeless_.
- Select an answer other than _Not at all_ for the question _Thoughts that you would be better off dead, or thoughts of hurting yourself in some way_.
- Save the Form.
- Open the _Visit information_ Form for editing once again.
- Select _Yes_ for the question _The patient has submitted the pre-appointment surveys?_
- Save the Form
- The alert email should be sent to the SMTP debugging console.
- Repeating these steps for a patient who as responded _Not at all_ for the question _Thoughts that you would be better off dead, or thoughts of hurting yourself in some way_ should not trigger this alert.
